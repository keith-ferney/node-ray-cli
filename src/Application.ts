import { Command } from './commands/Command';
import { Send } from './commands/Send';

export class Application {
    public y;
    public client: any;

    constructor(yargs: any = null, client: any = null, uuid: string | null = null) {
        this.y = yargs ?? require('yargs')(process.argv.slice(2));
        this.client = client;
        try {
            this.client.uuid = uuid;
        } catch (err) {
            //
        }
    }

    public commandClass = (...cmdClasses: any) => {
        cmdClasses.forEach(instance => {
            instance.client = this.client;
            this.y = this.y.command(instance.command, instance.help, instance.builder, instance.handle);
        });

        return this;
    };

    public alias(name: string, alias: string) {
        this.y = this.y.alias(name, alias);

        return this;
    }

    public help(option: string | null = null) {
        this.y = this.y.help(option);

        return this;
    }

    get argv() {
        return this.y.argv;
    }

    public run(commands: Command[]) {
        this.y = this.y.scriptName('ray');
        // @ts-ignore
        this.y = this.y.version(__APP_VERSION__); // eslint-disable-line no-undef

        commands.forEach(command => {
            command.client = this.client;

            try {
                command.uuid = this.client.uuid;
            } catch (err) {
                //
            }

            this.commandClass(command);
        });

        this.help('h').alias('h', 'help');

        const argv = this.argv;

        if (argv['_'].length === 1) {
            const commandNames = commands.map(command => command.name());

            if (!commandNames.includes(argv['_'][0] ?? '')) {
                argv['data'] = argv['_'][0];

                const cmd = new Send();

                cmd.client = this.client;
                cmd.uuid = this.client?.uuid ?? null;
                cmd.handle(argv);
            }

            return;
        }

        if (process.env.NODE_ENV !== 'test') {
            this.y.showHelp();
        }
    }
}
