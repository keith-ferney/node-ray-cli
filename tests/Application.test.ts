/* eslint-disable no-unused-vars */
/* eslint-disable no-undef */

import { Application } from '@/Application';
import { FakeYargs } from './TestClasses/FakeYargs';

let yargs: FakeYargs;
let app: Application;

beforeEach(() => {
    yargs = new FakeYargs();
    app = new Application(yargs);
});

it('creates aliases', () => {
    app.alias('v', 'verbose');
    app.alias('V', 'version');

    expect(yargs.aliases).toHaveLength(2);
    expect(yargs.aliases).toMatchSnapshot();
});

it('adds commands', () => {
    const cmds = [
        { command: 'one', help: 'test one ', builder: () => ({}), handle: args => ({}) },
        { command: 'two', help: 'test two ', builder: () => ({}), handle: args => ({}) },
    ];

    cmds.forEach(cmd => app.commandClass(cmd));

    expect(yargs.commandList).toHaveLength(cmds.length);
    expect(yargs.commandList).toMatchSnapshot();
});
