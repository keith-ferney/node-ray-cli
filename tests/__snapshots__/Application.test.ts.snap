// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds commands 1`] = `
[
  {
    "description": "test one ",
    "name": "one",
    "options": [Function],
  },
  {
    "description": "test two ",
    "name": "two",
    "options": [Function],
  },
]
`;

exports[`creates aliases 1`] = `
[
  {
    "alias": "verbose",
    "name": "v",
  },
  {
    "alias": "version",
    "name": "V",
  },
]
`;

exports[`runs the application 1`] = `
[
  {
    "description": "change to color of a previously sent payload",
    "name": "color <uuid> <color>",
    "options": [Function],
  },
]
`;

exports[`runs the application 2`] = `
[
  {
    "alias": "help",
    "name": "h",
  },
]
`;

exports[`runs the application 3`] = `"h"`;

exports[`runs the application and uses send as the default command 1`] = `
[
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "values": [
            "test message",
          ],
        },
        "origin": {
          "file": "/src/commands/Send.ts",
          "function_name": "xxxx",
          "hostname": "fake-host",
          "line_number": 999,
        },
        "type": "log",
      },
    ],
    "uuid": "fakeUuid",
  },
]
`;
